// trpc-nuxt/dist/client/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/client/index.ts
var client_exports = {};
__export(client_exports, {
  createTRPCNuxtClient: () => createTRPCNuxtClient,
  getQueryKey: () => getQueryKey,
  httpBatchLink: () => httpBatchLink,
  httpLink: () => httpLink
});
module.exports = __toCommonJS(client_exports);
var import_client2 = require("@trpc/client");
var import_shared2 = require("@trpc/server/shared");

// src/client/getQueryKey.ts
var import_ohash = require("ohash");
function getQueryKey(..._params) {
  const [procedure, input] = _params;
  if (typeof procedure === "string") {
    return getQueryKeyInternal(procedure, input);
  }
  const path = procedure._def().path;
  const dotPath = path.join(".");
  return getQueryKeyInternal(dotPath, input);
}
function getQueryKeyInternal(path, input) {
  return input === void 0 ? path : `${path}-${(0, import_ohash.hash)(input || "")}`;
}

// src/client/decorationProxy.ts
var import_shared = require("@trpc/server/shared");
var import_imports = require("#imports");
function isRefOrGetter(val) {
  return (0, import_imports.isRef)(val) || typeof val === "function";
}
function createAbortController(trpc) {
  let controller;
  if (trpc?.abortOnUnmount) {
    if ((0, import_imports.getCurrentInstance)()) {
      (0, import_imports.onScopeDispose)(() => {
        controller?.abort?.();
      });
    }
    controller = typeof AbortController !== "undefined" ? new AbortController() : {};
  }
  return controller;
}
function createNuxtProxyDecoration(name, client) {
  return (0, import_shared.createRecursiveProxy)((opts) => {
    const args = opts.args;
    const pathCopy = [name, ...opts.path];
    const lastArg = pathCopy.pop();
    const path = pathCopy.join(".");
    const [input, otherOptions] = args;
    if (lastArg === "_def") {
      return {
        path: pathCopy
      };
    }
    if (["useQuery", "useLazyQuery"].includes(lastArg)) {
      const { trpc, queryKey: customQueryKey, ...asyncDataOptions } = otherOptions || {};
      const controller = createAbortController(trpc);
      const queryKey = customQueryKey || getQueryKeyInternal(path, (0, import_imports.toValue)(input));
      const watch = asyncDataOptions.watch === false ? [] : isRefOrGetter(input) ? [...asyncDataOptions.watch || [], input] : asyncDataOptions.watch;
      const isLazy = lastArg === "useLazyQuery" ? true : asyncDataOptions.lazy || false;
      return (0, import_imports.useAsyncData)(queryKey, () => client[path].query((0, import_imports.toValue)(input), {
        signal: controller?.signal,
        ...trpc
      }), {
        ...asyncDataOptions,
        watch,
        lazy: isLazy
      });
    }
    if (lastArg === "useMutation") {
      const { trpc, ...asyncDataOptions } = otherOptions || {};
      const payload = (0, import_imports.ref)(null);
      const controller = createAbortController(trpc);
      const asyncData = (0, import_imports.useAsyncData)(() => client[path].mutate(payload.value, {
        signal: controller?.signal,
        ...trpc
      }), {
        ...asyncDataOptions,
        immediate: false
      });
      async function mutate(input2) {
        payload.value = input2;
        await asyncData.execute();
        return (0, import_imports.toRaw)(asyncData.data.value);
      }
      Object.assign(asyncData, { mutate });
      return asyncData;
    }
    return client[path][lastArg](...args);
  });
}

// src/client/links.ts
var import_client = require("@trpc/client");
var import_imports2 = require("#imports");
function isFetchError(error) {
  return error instanceof Error && error.name === "FetchError";
}
function customFetch(input, init) {
  return globalThis.$fetch.raw(input.toString(), init).catch((e) => {
    if (isFetchError(e) && e.response) {
      return e.response;
    }
    throw e;
  }).then((response) => ({
    ...response,
    headers: response.headers,
    json: () => Promise.resolve(response._data)
  }));
}
function httpLink(opts) {
  const headers = (0, import_imports2.useRequestHeaders)(opts?.pickHeaders);
  return (0, import_client.httpLink)({
    url: "/api/trpc",
    headers() {
      return headers;
    },
    fetch: customFetch,
    ...opts
  });
}
function httpBatchLink(opts) {
  const headers = (0, import_imports2.useRequestHeaders)(opts?.pickHeaders);
  return (0, import_client.httpBatchLink)({
    url: "/api/trpc",
    headers() {
      return headers;
    },
    fetch: customFetch,
    ...opts
  });
}

// src/client/index.ts
function createTRPCNuxtClient(opts) {
  const client = (0, import_client2.createTRPCProxyClient)(opts);
  const decoratedClient = (0, import_shared2.createFlatProxy)((key) => {
    return createNuxtProxyDecoration(key, client);
  });
  return decoratedClient;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createTRPCNuxtClient,
  getQueryKey,
  httpBatchLink,
  httpLink
});
